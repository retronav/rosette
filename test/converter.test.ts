import * as fs from "node:fs";
import * as path from "node:path";
import type { BlockObjectResponse, Client } from "@notionhq/client";
import { beforeAll, describe, expect, it, vi } from "vitest";
import { type NotionBlock, NotionConverter } from "../src/index.js";
import { createMockNotionClient } from "./mocks.js";

const baseFixturesDir = path.join(__dirname, "fixtures", "NotionConverter");
const blocksDir = path.join(baseFixturesDir, "blocks");
const edgeCasesDir = path.join(baseFixturesDir, "edge-cases");

describe("NotionConverter", () => {
	const blockMap = new Map<string, BlockObjectResponse>();
	const blockChildrenIndex = new Map<string, string[]>();

	beforeAll(() => {
		const blockFiles = fs.readdirSync(blocksDir);
		for (const file of blockFiles) {
			const filePath = path.join(blocksDir, file);
			const block: BlockObjectResponse = JSON.parse(
				fs.readFileSync(filePath, "utf-8")
			);
			blockMap.set(block.id, block);
			let parentId: string | undefined;
			if (block.parent) {
				if (block.parent.type === "block_id" && block.parent.block_id) {
					parentId = block.parent.block_id;
				} else if (block.parent.type === "page_id" && block.parent.page_id) {
					parentId = block.parent.page_id;
				}
			}

			if (parentId) {
				if (!blockChildrenIndex.has(parentId)) {
					blockChildrenIndex.set(parentId, []);
				}
				blockChildrenIndex.get(parentId)?.push(block.id);
			}
		}

		// Load edge case fixtures
		if (fs.existsSync(edgeCasesDir)) {
			const edgeCaseFiles = fs.readdirSync(edgeCasesDir);
			for (const file of edgeCaseFiles) {
				const filePath = path.join(edgeCasesDir, file);
				const block: BlockObjectResponse = JSON.parse(
					fs.readFileSync(filePath, "utf-8")
				);
				blockMap.set(block.id, block);
			}
		}
	});

	const mockClient = createMockNotionClient(blockMap, blockChildrenIndex);
	const converter = new NotionConverter(mockClient);

	it("converts blocks to HTML", async () => {
		const pageData = await converter.fetch(
			"1f6aa550-f576-802c-ab65-e3e8c49a8bcd"
		);
		// Extract the children blocks from the page data
		const blocks = pageData.children;
		const html = converter.blocksToHtml(blocks);

		expect(html).toMatchSnapshot();
	});

	describe("Error Handling", () => {
		it("should throw error for invalid block type during HTML conversion", () => {
			const invalidBlock = JSON.parse(
				fs.readFileSync(
					path.join(edgeCasesDir, "invalid-block-type.json"),
					"utf-8"
				)
			) as NotionBlock;

			expect(() => {
				converter.blocksToHtml([invalidBlock]);
			}).toThrow(
				/Failed to parse block: invalid-block-123 \(unsupported_block_type\)/
			);
		});

		it("should throw error for malformed block structure", () => {
			const malformedBlock = JSON.parse(
				fs.readFileSync(
					path.join(edgeCasesDir, "malformed-paragraph.json"),
					"utf-8"
				)
			) as NotionBlock;

			expect(() => {
				converter.blocksToHtml([malformedBlock]);
			}).toThrow(
				/Failed to parse block: malformed-paragraph-456 \(paragraph\)/
			);
		});

		it("should handle network errors during fetch", async () => {
			const mockFailingClient = {
				blocks: {
					retrieve: vi.fn().mockRejectedValue(new Error("Network timeout")),
					children: {
						list: vi.fn().mockRejectedValue(new Error("Network error"))
					}
				}
			} as unknown as Client;

			const failingConverter = new NotionConverter(mockFailingClient);

			await expect(
				failingConverter.fetch("nonexistent-block-id")
			).rejects.toThrow("Network timeout");
		});

		it("should handle missing block references", async () => {
			const mockClientWithMissingBlocks = {
				blocks: {
					retrieve: vi.fn().mockRejectedValue(new Error("Block not found")),
					children: {
						list: vi.fn().mockResolvedValue({ results: [] })
					}
				}
			} as unknown as Client;

			const converter = new NotionConverter(mockClientWithMissingBlocks);

			await expect(converter.fetch("missing-block-id")).rejects.toThrow(
				"Block not found"
			);
		});

		it("should handle empty blocks array gracefully", () => {
			const html = converter.blocksToHtml([]);
			expect(html).toBe("");
		});

		it("should handle corrupted children data during fetch", async () => {
			const mockClientWithCorruptedChildren = {
				blocks: {
					retrieve: vi.fn().mockResolvedValue({
						id: "test-block",
						type: "paragraph",
						has_children: true,
						paragraph: { rich_text: [] }
					}),
					children: {
						list: vi.fn().mockResolvedValue({
							results: [
								{
									id: "child-1",
									type: "paragraph",
									paragraph: { rich_text: [] }
								}
							]
						})
					}
				}
			} as unknown as Client;

			const converter = new NotionConverter(mockClientWithCorruptedChildren);

			// Mock the second retrieve call to fail
			vi.mocked(mockClientWithCorruptedChildren.blocks.retrieve)
				.mockResolvedValueOnce({
					id: "test-block",
					type: "paragraph",
					has_children: true,
					// @ts-expect-error: corrupted data
					paragraph: { rich_text: [] }
				})
				.mockRejectedValueOnce(new Error("Child block fetch failed"));

			await expect(converter.fetch("test-block")).rejects.toThrow(
				"Child block fetch failed"
			);
		});

		it("should handle toggle blocks with missing rich_text", () => {
			const brokenToggleBlock = {
				id: "broken-toggle",
				type: "toggle",
				toggle: {
					// Missing rich_text property
					color: "default"
				}
			} as NotionBlock;

			expect(() => {
				converter.blocksToHtml([brokenToggleBlock]);
			}).toThrow(/Failed to parse block: broken-toggle \(toggle\)/);
		});
	});
});
