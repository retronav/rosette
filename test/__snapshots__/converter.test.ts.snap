// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NotionConverter > converts blocks to HTML 1`] = `"<p class="">some text</p><details><summary class="">What is love?</summary><div><p class="">Baby don’t hurt me…</p></div></details><ul><li class="">Oranges</li></ul><p class="">some text</p><ol><li class="">Lakers</li></ol><p class="">some text</p><p class=""></p><ul><li class="">Apples</li></ul><ol><li class="">Warriors</li></ol><ol><li class="">Mavericks</li></ol><p class=""></p><p class=""><strong>Bold </strong><em><strong>Italic </strong></em><u><em><strong>Underline </strong></em></u><u><del><em><strong>StrikeThrough</strong></em></del></u><u><em><strong> </strong></em></u><code><u><em><strong>Code </strong></em></u></code><span class="notion-yellow-background"><code><u><em><strong>Color</strong></em></u></code></span></p><p class=""></p><h1>Heading 1</h1><figure><video src="https://www.youtube.com/watch?v=dQw4w9WgXcQ" controls></video><figcaption></figcaption></figure><p class=""></p><p class=""></p><blockquote class=""><em><strong>“One must have chaos in one's self to give birth to a dancing star</strong></em><strong>.”</strong> -Nietzsche</blockquote><ul><li class=""><input type="checkbox" checked>Eat apples</li></ul><p class=""></p><figure><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/2/20/YouTube_2024.svg/500px-YouTube_2024.svg.png"><figcaption>Youtube Logo</figcaption></figure><figure><img src="https://prod-files-secure.s3.us-west-2.amazonaws.com/4c6329c4-ddf4-4a16-8065-9bb204231c5d/32632a86-ac65-4998-a785-446e7160c31b/logo.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&#x26;X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&#x26;X-Amz-Credential=ASIAZI2LB4663CGSEMHU%2F20250523%2Fus-west-2%2Fs3%2Faws4_request&#x26;X-Amz-Date=20250523T174001Z&#x26;X-Amz-Expires=3600&#x26;X-Amz-Security-Token=IQoJb3JpZ2luX2VjEDkaCXVzLXdlc3QtMiJHMEUCIQCuB9u5qzGgFW1D%2B9KvBYSaRr81W%2BlTmqfputF6Y%2B3FAQIgLYY1i4NdUU1fqA5C7ugpKKOADCEQZXauVNPrpx4Sk4UqiAQI8v%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FARAAGgw2Mzc0MjMxODM4MDUiDHeY0hpjgdiaD0zlmircA4AQx6JHNL7UKkRDKqtqLprQndavAvwBfh3G0Gd85Y8LPwkGuMiaDcGyQt4J1HZKCxkQqef2eguB8RbR1jRKpdopJmCK7krkCNh05%2BNe08q9p6dyRAYldy52G3bkcCwxRTs8TkgGuaJUEo3%2BPCuck4dQGRbtq2UAEgKEvUp0mbCgwOl5rAs1GcFcVCRqSK7XSoPSSoKAJjlFBhwHfUYZ0ScwdT7qIxtQRQ2T00Bpx4U%2BVD%2FvYe3bI5ezIndJGptTIiy26P%2FBAvI53PfIUhsJF9GZwW6pykmKXmXoqtnrRmJQxcZ1W%2FiUHxFVyNiQ28Wv7CoFdNpXbT6H1ZojjPEfHLXQiQgSdMewS4czlMKc2pjzXv2bMp3k%2Ft4f4mWf4gwVLKDDW3zbY2QRXItxjobc1DW33qUfKQxJpmTekW9Vz8fX9jlNUB23kWpKdDhYxcQWPZM0YjqYc4kC8G2FKjohS0qMb%2FhEcgA43vSdCQP3QLNctEW0yyzzuhZeM5v0iQPrDv44poyUJEjBZO%2Bkq7ny9FNdZ9U%2Bhu6%2FohgYSOk7teQcBZKX%2B%2BMfjGM4JD%2FUZwFfDUmmQQrCo3Yy6G03iexVjH2%2BtzsVclewfBjBqPFxbEyNl1SG1eMEVNORrVwSMIbWwsEGOqUBUUO%2FkNSipcQHkVs3%2BoOM%2BNnc52dtdw5Opsll5Kx5mz%2FW3541WMTRKzAd%2FG5OqFpqi2NWgWSveJfHNbVBxufX67LQ4%2B5oXgX%2BcbmWdWY%2FCWUAQhHLDF03KbpapNXAd0V3HxwXd8nIzXlU%2BiIcUMHW%2FnNhV9nPXtxDIGT1VBeMDdMNKTc%2B9m%2BO7H1eMeX%2Fo80lnPlL9HL3BSIE6jzhXIMTSbq0u0f7&#x26;X-Amz-Signature=08e0df08add427276a4285c37887cc4c540e98d76ddcb2fa6b49c01a49c04704&#x26;X-Amz-SignedHeaders=host&#x26;x-id=GetObject"><figcaption>Rosette logo</figcaption></figure><table><tbody><tr><td>Sr. No</td><td>Name</td><td>Apples</td></tr><tr><td>2</td><td>John</td><td><em><strong>10</strong></em></td></tr><tr><td>1</td><td>Alice</td><td><em><strong>6</strong></em></td></tr></tbody></table><p class=""><a data-notion-page-mention="126744b6-a7b9-433a-bf8a-78db40eb004e">Untitled</a> </p><h3>Heading 3</h3><code class="language-math">x^n + y^n = z^n</code><ul><li class="">Mangoes</li></ul><p class=""></p><h2>Heading 2</h2><pre><code class="language-javascript">console.log("Hello World!");</code></pre><p class=""></p><p class=""></p><p class=""></p><p class=""><code class="language-math">x = y + z</code></p>"`;
